concepts/pipeline/data_pipeline_detail.md

```diff
@@ -0,0 +1,429 @@
+
+There is a separate data store for self-serve **Analysis Outputs**, intended to keep ad-hoc, temporary data out of the Data Lake. This is implemented as a separate S3 location, with personal output locations prefixed with each person's user id, similar to the layout of the `/home` directory on a Unix system.
+
+Analysis outputs can also be made public using the **Public Outputs** bucket. This is a web-accessible S3 location for powering public dashboards.
+
+## Stream Processing
+
+Stream processing is done using [Hindsight] and [Spark Streaming].
+
+Hindsight allows you to run [plugins written in Lua inside a sandbox][lua_sandbox]. This gives a safe, performant way to do self-serve streaming analysis. See [this article](cookbooks/realtime_analysis_plugin.md) for an introduction. Hindsight plugins do the initial data validation and decoding, as well as writing out to long-term storage in both [Heka-framed protobuf][heka_protobuf] and [parquet][direct2parquet] forms.
+
+Spark Streaming is used to read from Kafka and perform [low-latency ETL and aggregation tasks][telemetry-streaming]. These aggregates are currently used by [Mission Control] and are also available for querying via [Re:dash].
+
+## Batch Processing
+
+Batch processing is done using [Spark]. Production ETL code is written in both [Python][python_mozetl] and [Scala][telemetry-batch-view].
+
+There are [Python][py_dataset] and [Scala][moztelemetry] libraries for reading data from the Data Lake in [Heka-framed protobuf][heka_protobuf] form, though it is much easier and more performant to make use of a [derived dataset](/concepts/choosing_a_dataset.md) whenever possible.
+
+Datasets in parquet format can be read natively by Spark, either using Spark SQL or by reading data directly from S3.
+
+Data produced by production jobs go into the Data Lake, while output from ad-hoc jobs go into Analysis Outputs.
+
+Job scheduling and dependency management is done using [Airflow]. Most jobs run once a day, processing data from "yesterday" on each run. A typical job launches a cluster, which fetches the specified ETL code as part of its bootstrap on startup, runs the ETL code, then shuts down upon completion. If something goes wrong, a job may time out or fail, and in this case it is retried automatically.
+
+## Self Serve Data Analysis
+
+```mermaid
+graph TD
+  subgraph Storage
+    lake[Data Lake]
+    s3_output_public[Public Outputs]
+    s3_output_private[Analysis Outputs]
+  end
+  subgraph ATMO
+    Jupyter -->|runs on| emr_cluster[Ad hoc EMR Cluster]
+    Zeppelin -->|runs on| emr_cluster
+    atmo_service[ATMO Service] -->|launch| emr_cluster
+    atmo_service -->|schedule| emr_job[fa:fa-clock-o Scheduled EMR Job]
+    emr_cluster -->|mount| EFS
+    emr_cluster --> lake
+    emr_job --> s3_output_public
+    emr_job --> s3_output_private
+  end
+  subgraph STMO
+    redash[Re:dash] -->|Presto| lake
```

Comment:
This arrow's direction confused me on first read. Maybe it makes sense to annotate the relationship? E.g. |Presto Consumes|?

The same is true for the relationship between Databricks, Ad-hoc clusters, and Scheduled jobs. Are they consuming or producing data from the datalake / public buckets?

Reply:
Ah yes, I can see the confusion here. The arrows on this graph have mixed meaning - either "reads from" or "writes to".

Reply:
I've changed these arrows to be more explicit in terms of reading and writing. I think the fact that the Presto engine is used is covered sufficiently elsewhere, and this makes the data flow easier to follow.

Reply:
Looks great, much more clear to me. Thanks!